### PR TITLE
Changed default for histogram bins from closed-right to closed-left

### PR DESCRIPTION
**Pull request by oschulz**
See #184.

### CODE REVIEW

**Comment by nalimilan** [comment]
(Travis failure is unrelated.)

**Comment by nalimilan** [comment]
I think we should print a deprecation warning (`Base.depwarn`) when `closed` isn't specified, explaining that the default changed. Else people will get incorrect results without notice.

**Comment by oschulz** [comment]
> I think we should print a deprecation warning

But how, exactly - Keyword arguments have to have a default value, after all. Is there a "standard way" to do this? Or do we just use `closed = :default_left` or so, and depwarn if it's not replaced by `:left`  or `right` by the user? The risk here is that users will see that `:default_left` in the signature, and start using it in their code.

The other thing is what we'll have to change a lot of the test code, and change it back later on, when we disable the deprecation (I assume it's only intended to stay for a limited time?).

**Comment by nalimilan** [comment]
`default_left` sounds OK. People won't use it because 1) you won't show it in the docs at all, so they won't see it, and 2) it will print an annoying warning. I know that's painful, but changing behaviour silently is really too dangerous.

**Comment by oschulz** [comment]
> :default_left sounds OK

@nalimilan: Ok, I'll get on it. I'll put in some comments to make it easier to remove afterwards.

**Comment by oschulz** [comment]
Ok, deprecation warnings are in. Maybe we shouldn't leave them in for too long ... ;-)

**Comment by oschulz** [comment]
Forgot deprecation handling in two places, fixed now.

File: src/hist.jl [hunk 1]
```diff
@@ -97,20 +109,21 @@ type Histogram{T<:Real,N,E} <: AbstractHistogram{T,N,E}
     weights::Array{T,N}
     closed::Symbol
     function Histogram(edges::NTuple{N,AbstractArray},weights::Array{T,N},closed::Symbol)
+
```
Comment (nalimilan) [inline src/hist.jl]: Please remove that unrelated line break addition.

File: test/hist.jl [hunk 2]
```diff
@@ -2,28 +2,28 @@ using StatsBase
...
-@test sum(fit(Histogram,[1,2,3]).weights) == 3
-@test fit(Histogram,Int[]).weights == Int[]
-@test fit(Histogram,[1]).weights == [1]
-@test fit(Histogram,[1,2,3],[0,2,4]) == Histogram([0,2,4],[2,1])
-@test fit(Histogram,[1,2,3],[0,2,4]) != Histogram([0,2,4],[1,1])
-@test fit(Histogram,[1,2,3],0:2:4) == Histogram(0:2:4,[2,1])
-@test all(fit(Histogram,[1:100;]/100,0.0:0.01:1.0).weights .==1)
-@test fit(Histogram,[1,1,1,1,1]).weights[1] == 5
-@test sum(fit(Histogram,(rand(100),rand(100))).weights) == 100
+@test sum(fit(Histogram,[1,2,3], closed=:left).weights) == 3  # Explicit arg "closed" during default deprecation
```
Comment (nalimilan) [inline test/hist.jl]: Looks too verbose to me. Could you simply add a `FIXME: closed` comment at the beginning of the block? A good strategy would be to add a comment at `_check_closed_arg` saying to adapt all places marked `FIXME: closed` when removing that warning, and simply marking the relevant places that way.

Comment (oschulz) [inline test/hist.jl]: Will do.

File: src/hist.jl (hunk 1 above)
Comment (oschulz) [inline src/hist.jl]: Ooops, sorry, that somehow snuck in there ...

File: test/hist.jl (hunk 2 above)
Comment (oschulz) [inline test/hist.jl]: Ok, done. I rebased the PR branch to keep history neat.

**Comment by oschulz** [comment]
@nalimilan, are you happy with it this way?

**Comment by andreasnoack** [comment]
@oschulz @nalimilan is offline the next nine days so it might take a while before he can review and I think we should wait for his comments.

**Comment by oschulz** [comment]
@andreasnoack: Thanks for the heads-up. Well, a few days won't matter, people will still have to write `closed = :left` during the deprecation period anyhow. :-)

**Comment by oschulz** [comment]
I lost track of this - I thought it was merged already. I just rebased on the current master - could someone review (and hopefully merge)?

**Comment by oschulz** [comment]
@ararslan - while you're in review mode, could you take a look at this, too? This PR was pretty much agreed on, but then all of us just kinda lost track of it.

File: test/hist.jl [hunk 3]
```diff
@@ -2,28 +2,31 @@ using StatsBase
...
+@test fit(Histogram,[1,1,1,1,1], closed=:left).weights[1] == 5  # FIXME: closed
+@test sum(fit(Histogram,(rand(100),rand(100)), closed=:left).weights) == 100  # FIXME: closed
 @test fit(Histogram,1:100,nbins=5,closed=:right).weights == [20,20,20,20,20]
 @test fit(Histogram,1:100,nbins=5,closed=:left).weights == [19,20,20,20,20,1]
 @test fit(Histogram,0:99,nbins=5,closed=:right).weights == [1,20,20,20,20,19]
 @test fit(Histogram,0:99,nbins=5,closed=:left).weights == [20,20,20,20,20]
 
-@test fit(Histogram,(1:100,1:100),nbins=5).weights == diagm([20,20,20,20,20])
-@test fit(Histogram,(1:100,1:100),nbins=(5,5)).weights == diagm([20,20,20,20,20])
+# FIXME: closed (all lines in this block):
```
Comment (ararslan) [inline test/hist.jl]: What's meant by `FIXME: closed` here and elsewhere?

Comment (oschulz) [inline test/hist.jl]: I don't remember who came up with it - these are meant to be a reminder to remove `closed = :left` and similar from the code when the deprecation period (during which closed *has* to be specified, so that people will notice the changed default) is over.

Comment (ararslan) [inline test/hist.jl]: Ah, that makes sense. It would be good to make the comment more explicit then.

**Comment by oschulz** [comment]
Closed #184.

File: test/hist.jl (hunk 3 above)
Comment (oschulz) [inline test/hist.jl]: Sure - what would you suggest? Maybe just a comment on top of the file, stating

    # FIXME: closed - comments imply that `closed = :left` arguments should be removed after
    # the deprecation period on closed (following default change from :right to :left) has ended.

Or something like that? Whatever you think will be clear to the maintainers.

Comment (ararslan) [inline test/hist.jl]: Sure, that works for me.

Comment (oschulz) [inline test/hist.jl]: Ah, now I that I look at it again - there's actually a comment on top of hist.jl that explains it in detail. :-)

Comment (oschulz) [inline test/hist.jl]: New commit - Ok like that?

Comment (ararslan) [inline test/hist.jl]: So there is! Sorry, I completely missed that. What you have here is fine then.

Comment (ararslan) [inline test/hist.jl]: > Ok like that?

Yes indeed. Thanks!

**Comment by oschulz** [comment]
Hey, can I just ask - what's the status on this, concerning merging?

**Comment by ararslan** [comment]
Thanks for the reminder!
